### PR TITLE
Brimdemons & Lobstrosities drop (slightly) useful organs

### PR DESCRIPTION
**Pull request by Jacquerel**
## About The Pull Request

Goliaths, Legions, Watchers, and (as of recently) Bileworms all drop something vaguely useful when they die.
Brimdemons and Lobstrosities do not. This PR aims to fix that, so that there's at least some vague benefit to hunting them.

In this case it takes the form of organs you get when you butcher them, similar to the regenerative core from Legions.
As they're similar to the regenerative core, I modified the regenerative core to extend from a new common "monster core" typepath which these two new organs also extend.
Like the regenerative core, both of these items do something when used and something slightly different if you go to the effort of having someone implant them into your body. They also decay over time, and you can use stabilising serum to prevent this from happening.

https://user-images.githubusercontent.com/7483112/195967746-55a7d04d-224e-412d-aedc-3a0ec754db3d.mp4

The Rush Gland from the Lobstrosity lets you do a little impression of their charging attack, making you run very fast for a handful of seconds and ignoring slowdown effects. Unlike a lobstrosity you aren't actually built to do this so if you run into a mob you will fall over, and if you are doing this on the space station running into any dense object will also make you fall over (it shouldn't make you _too_ much of a pain for security to catch).
The idea here is that you use this to save time running back and forth from the mining base.

The Brimdust Sac from the Brimdemon covers you in exploding dust. The next three times you take Brute damage some of the dust will explode, dealing damage equal to an unupgraded PKA shot to anything near you (but not you).
If you do this in a space station not only is the damage proportionally lower (still matching the PKA), but it _does_ effect you and also it sets you on fire. You can remove the buff by showering it off.
The idea here is that you use this for minor revenge damage on enemies whose attacks you don't manage to dodge.

https://user-images.githubusercontent.com/7483112/195967811-0b362ba9-2da0-42ac-bd55-3809473cbc74.mp4

If you implant the Rush Gland then you can use it once every 3 minutes without consuming it, and the buff lasts very slightly longer. It will automatically trigger itself if your health gets low, which might be good (helps you escape a rough situation) or bad (didn't want to use it yet).

https://user-images.githubusercontent.com/7483112/195967888-f63f7cbd-60cd-4309-8004-203afc5b2153.mp4

If you implant the Brimdust Sac then you can use it once every 3 minutes to shake off cloud of dust which gives the buff to everyone nearby, if you want to kit out your miner squad. The dust cloud also makes you cough if you stand in it, and it's opaque. If you catch fire with this organ inside you and aren't in mining atmosphere then it will explode inside of your abdomen, which should probably be avoided, resultingly it is very risky to use this on the space station.

## Why It's Good For The Game

There's a small aspect of the Shaft Miner job where you're living off the land and scavenging parts off creatures you have hunted and I think that's pretty cool, so I would like some more of it.
It is also nice to mitigate the slight disappointment of killing a monster which just drops raw meat and maybe a pile of reagents for botany which you can't even pick up unless you brought a beaker with you for some reason.
I don't think people will really regularly be having these surgically implanted into them but maybe someone can think of something funny to do with a cloud of dust which makes people explode when they get punched some day.
Ashwalker content...?

## Changelog

:cl:
add: You can harvest useful organs from Brimdemons and Lobstrosities.
fix: You can no longer 'preserve' regenerative cores which have already expired to repair them.
/:cl:

### CODE REVIEW

File: code/modules/mining/equipment/monster_organs/monster_organ.dm [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+/// Set up the action for using the organ internally
+/obj/item/organ/internal/monster_core/proc/setup_internal_use_action()
+	use_internal = new(src)
+
+/obj/item/organ/internal/monster_core/Destroy(force, silent)
+	QDEL_NULL(use_internal)
+	return ..()
+
+/obj/item/organ/internal/monster_core/Insert(mob/living/carbon/target_carbon, special = 0, drop_if_replaced = TRUE)
+	. = ..()
```
Comment (Kapu1178) [inline code/modules/mining/equipment/monster_organs/monster_organ.dm:79]: pretty sure you need an `if(!.) return` here, dont remember off the top of my head

File: code/modules/mining/equipment/monster_organs/monster_organ.dm [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+/obj/item/organ/internal/monster_core/afterattack(atom/target, mob/user, proximity_flag)
+	. = ..()
+	if (!proximity_flag)
+		return
+	try_apply(target, user)
+
+/obj/item/organ/internal/monster_core/attack_self(mob/user)
+	if (!user.canUseTopic(src, be_close = TRUE, no_dexterity = FALSE, no_tk = TRUE))
+		return
+	try_apply(user, user)
```
Comment (Kapu1178) [inline code/modules/mining/equipment/monster_organs/monster_organ.dm:150]: you know, at some point you gotta ask, why is this even an organ.

File: code/modules/mining/equipment/monster_organs/monster_organ.dm [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	return TRUE
+
+/**
+ * Decays the organ, it is now useless.
+ */
+/obj/item/organ/internal/monster_core/proc/go_inert()
+	if (inert)
+		return FALSE
+	inert = TRUE
+	decay_timer = null
```
Comment (Kapu1178) [inline code/modules/mining/equipment/monster_organs/monster_organ.dm:116]: deltimer?

File: code/modules/mining/equipment/monster_organs/regenerative_core.dm [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		return
+	SSblackbox.record_feedback("nested tally", "hivelord_core", 1, list("[type]", "inert"))
+
+/obj/item/organ/internal/monster_core/regenerative_core/on_life(delta_time, times_fired)
+	. = ..()
+	if (owner.health <= owner.crit_threshold)
+		activate_implanted()
+
+/obj/item/organ/internal/monster_core/regenerative_core/activate_implanted()
+	owner.revive(full_heal = TRUE, admin_revive = FALSE)
```
Comment (Kapu1178) [inline code/modules/mining/equipment/monster_organs/regenerative_core.dm:32]: wow thats kinda busted

File: code/modules/mining/equipment/monster_organs/rush_gland.dm [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	name = "rush gland"
+	icon_state = "lobster_gland"
+	icon_state_preserved = "lobster_gland_stable"
+	icon_state_inert = "lobster_gland_decayed"
+	desc = "A lobstrosity's engorged adrenal gland. You can squeeze it to get a rush of energy on demand."
+	desc_preserved = "A lobstrosity's engorged adrenal gland. It is preserved, allowing you to use it for a burst of speed whenever you need it."
+	desc_inert = "A lobstrosity's adrenal gland. It is all shrivelled up."
+	user_status = /datum/status_effect/lobster_rush
+
+#define HEALTH_DANGER_ZONE 30
```
Comment (Kapu1178) [inline code/modules/mining/equipment/monster_organs/rush_gland.dm:15]: put defines at the file header and undef at footer, nesting it between procs is basically pointless.

File: code/modules/mining/equipment/monster_organs/rush_gland.dm [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	if (lavaland_equipment_pressure_check(get_turf(source)))
+		return
+	smack_into(source)
+	source.visible_message(span_warning("[source] crashes into [target]!"))
+	qdel(src)
+
+/datum/status_effect/lobster_rush/proc/smack_into(mob/living/target)
+	target.Knockdown(5 SECONDS)
+	target.adjustStaminaLoss(40)
+	target.adjustBruteLoss(20)
```
Comment (Kapu1178) [inline code/modules/mining/equipment/monster_organs/rush_gland.dm:82]:
```suggestion
	target.apply_damage(20, BRUTE)
```
adjustBruteLoss() ignores armor. There's other args to apply_damage you will want to use here, i do not remember the exact order.

spread_damage should be TRUE tho

File: code/modules/mining/lavaland/tendril_loot.dm [hunk 7]
```diff
@@ -624,7 +624,17 @@
...
+		/obj/item/flashlight,
+		/obj/item/tank/internals,
+		/obj/item/pickaxe,
+		/obj/item/spear,
+		/obj/item/organ/internal/monster_core,
+		/obj/item/knife,
+		/obj/item/kinetic_crusher,
+		/obj/item/resonator,
+		/obj/item/melee/cleaving_saw,
+		)
```
Comment (Kapu1178) [inline code/modules/mining/lavaland/tendril_loot.dm:637]:
```suggestion
	allowed = list(
		/obj/item/flashlight,
		/obj/item/tank/internals,
		/obj/item/pickaxe,
		/obj/item/spear,
		/obj/item/organ/internal/monster_core,
		/obj/item/knife,
		/obj/item/kinetic_crusher,
		/obj/item/resonator,
		/obj/item/melee/cleaving_saw,
	)
```
formatting nitpick

File: code/modules/mining/equipment/monster_organs/monster_organ.dm [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+	/// Icon state to apply when inert
+	var/icon_state_inert
+	/// Description to use once organ is inert
+	var/desc_inert
+	/// Status effect applied by this organ
+	var/datum/status_effect/user_status
+	/// Moodlet applied by this organ
+	var/datum/mood_event/moodlet
+	/// Action to grant when organ is implanted
+	var/datum/action/item_action/organ_action/use_internal
```
Comment (Kapu1178) [inline code/modules/mining/equipment/monster_organs/monster_organ.dm:62]: If you're going to add this much boilerplate, don't add it to a random subtype. Preservation should be an `/organ/internal` mechanic, as should `moodlet on insert` and `action to grant`

**Review by Kapu1178 — CHANGES_REQUESTED** [review]

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Comment by rasepretep** [comment]
I like this not as a miner, but as someone who likes to play Ashwalker or go down to live as an Ashssistant. This'll make both of these a lot more fun.

File: code/modules/mining/equipment/monster_organs/brimdust_sac.dm [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+	id = "brimdust_coating"
+	stacks = 3
+	max_stacks = 3
+	tick_interval = -1
+	consumed_on_threshold = FALSE
+	alert_type = /atom/movable/screen/alert/status_effect/brimdust_coating
+	/// Damage to deal on explosion
+	var/static/blast_damage = 40
+	/// Damage reduction when not in a mining pressure area
+	var/static/pressure_modifier = 0.25
```
[the remark this answers is not in the document]
Comment (Jacquerel) [reply code/modules/mining/equipment/monster_organs/brimdust_sac.dm:82]: Honestly cannot tell you why I did that

File: code/modules/mining/equipment/monster_organs/monster_organ.dm (hunk 2 above)
Comment (Jacquerel) [reply code/modules/mining/equipment/monster_organs/monster_organ.dm:153]: You're not really wrong that probably the instances of people implanting even the existing item as an organ happens one round in a hundred (even though as you noticed, it provides a full admin heal!) as opposed to just using it inhand as soon as you see it on the floor
I think that for the sandbox it's still fun to have the option to see a weird alien organ and have the ability to ask a doctor to put it into your body though

**Comment by ATH1909** [comment]
rush gland boost should apply a positive modifier for tackles, imo

**Comment by JohnFulpWillard** [comment]
Does this have any purpose on lavaland or does it only benefit PvP? Being immediately sent to stamcrit mid-megafauna doesn't sound like something miners would want.

**Comment by Jacquerel** [comment]
> Does this have any purpose on lavaland or does it only benefit PvP? Being immediately sent to stamcrit mid-megafauna doesn't sound like something miners would want.

Moving faster is useful basically anywhere, especially if you're actually mining (I know this might be an unlikely suggestion). 
The speed boost isn't really meant to be used as an attack because it's as debilitating to you as your target and does (as you have identified) basically nothing to megafauna, but they pretty much all have a place where a sudden boost of speed would be convenient.
I can probably make it add a trait which makes tackles better for the very short period of time it is active though.

The retaliation damage is very dangerous to yourself on the space station as a way to try and make it less useful for PvP unless you're already fighting a miner in the place where they're strongest, whereas you're pretty likely to take incidental melee attacks from mining mobs every so often. Not particularly useful on megafauna because ideally you're not getting damaged by them at all but if you are, damage is damage.

**Comment by TheGreatValvatorez** [comment]
brim gland pls

**Comment by Shadowflame909** [comment]
Yeah getting stunned while trying to run away from a mega fauna (owie they hit you) doesn't sound too good and might end up being counterintuitive/unused like the cursed heart. It's well balanced for the station though!

**Comment by optimumtact** [comment]
> Yeah getting stunned while trying to run away from a mega fauna (owie they hit you) doesn't sound too good and might end up being counterintuitive/unused like the cursed heart. It's well balanced for the station though!

This encapsulates everything wrong with lavaland in a single comment lol

**Comment by termaus** [comment]
I dont understand, isnt this basically jump boots with major debuffs with slightly longer range? I think this item would be greatly unbalanced, a suggestion is significanly longer duration for lobstrosity gland with something like %5 chance to drop

**Review by MrMelbert — APPROVED** [review]
Code looks good

**Comment by Sealed101** [comment]
gonna leave this here

image

File: code/modules/mining/equipment/monster_organs/rush_gland.dm [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	desc_inert = "A lobstrosity's adrenal gland. It is all shrivelled up."
+	user_status = /datum/status_effect/lobster_rush
+	actions_types = list(/datum/action/cooldown/monster_core_action/adrenal_boost)
+
+/obj/item/organ/internal/monster_core/rush_gland/on_life(delta_time, times_fired)
+	. = ..()
+	if (owner.health <= HEALTH_DANGER_ZONE)
+		trigger_interal_action()
+
+/obj/item/organ/internal/monster_core/rush_gland/trigger_interal_action()
```
Comment (Fikou) [inline code/modules/mining/equipment/monster_organs/rush_gland.dm:24]: internal?

File: code/modules/mining/equipment/monster_organs/monster_organ.dm [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+ * Useful organs which drop as loot from a mining creature.
+ * Generalised behaviour is that they will decay and become useless unless provided with serum.
+ * These should usually do something both when used in-hand, or when implanted into someone.
+ */
+/obj/item/organ/internal/monster_core
+	name = "monster core"
+	desc = "All that remains of a monster. This abstract item should not spawn. \
+		It will rapidly decay into uselessness. but don't worry because it's already useless."
+	icon = 'icons/obj/medical/organs/mining_organs.dmi'
+	icon_state = "roro core 2"
```
Comment (Fikou) [inline code/modules/mining/equipment/monster_organs/monster_organ.dm:37]: rename this to hivelord_core, roros dont exist anymore

File: code/modules/mining/equipment/monster_organs/monster_organ.dm [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+
+/**
+ * Decays the organ, it is now useless.
+ */
+/obj/item/organ/internal/monster_core/proc/go_inert()
+	if (inert)
+		return FALSE
+	inert = TRUE
+	deltimer(decay_timer)
+	decay_timer = null
```
Comment (Fikou) [inline code/modules/mining/equipment/monster_organs/monster_organ.dm:110]: why are you setting the timer to null if you already deltimered it

**Comment by LarissaMayrink** [comment]
Please don't give miners a free regenerating meth implant from one of the weakest mobs on lavaland

File: code/modules/mining/equipment/monster_organs/rush_gland.dm (hunk 10 above)
Comment (Jacquerel) [reply code/modules/mining/equipment/monster_organs/rush_gland.dm:24]: The proc name is meant to differentiate "thing it does if you squeeze it" vs "thing it does if you have it inside your body".
But maybe "organ action" or something might be clearer.

**Comment by Jacquerel** [comment]
> Please don't give miners a free regenerating meth implant from one of the weakest mobs on lavaland

While I am happy to make changes, the implanted action is active for 5 seconds every 3 minutes and if you use it on the space station you have to be very careful not to end it early by running into something and immobilising yourself for just as long as the buff would have lasted (for instance, closed doors). 
And this only after you can convince someone to surgically insert it into your body, which certainly isn't impossible but is a sufficient risk that it is a deterrent to most people implanting other useful equipment.

**Comment by LarissaMayrink** [comment]
>And this only after you can convince someone to surgically insert it into your body

lrp players view of the game. Also even on bagil which is wilder than sybil you can get implants done easily, you just gotta be nice. 

I'm a sucker for teethpills and have gotten one on literally every bagil round i've played in the last two weeks, besides a lot of doctors are always taking any chance to do surgery

**Review by Fikou — CHANGES_REQUESTED** [review]
make icon names of hivelords and legions consistent with the other 2, and also set the hivelord's decaying icon

**Pull request merged**